src/uct/ib/mlx5/ib_mlx5_md.c

```diff
@@ -199,6 +199,57 @@ static ucs_mpool_ops_t uct_ib_mlx5_dbrec_ops = {
     .obj_cleanup   = NULL
 };
 
+static ucs_status_t uct_ib_mlx5dv_check_odp(uct_ib_mlx5_md_t *md, void *cap) {
+    uint32_t out[UCT_IB_MLX5DV_ST_SZ_DW(query_hca_cap_out)] = {};
+    uint32_t in[UCT_IB_MLX5DV_ST_SZ_DW(query_hca_cap_in)] = {};
+    void *odp;
+    int ret;
+
+    if (!UCT_IB_MLX5DV_GET(cmd_hca_cap, cap, pg)) {
+        return UCS_OK;
+    }
+
+    odp = UCT_IB_MLX5DV_ADDR_OF(query_hca_cap_out, out, capability);
+    UCT_IB_MLX5DV_SET(query_hca_cap_in, in, opcode, UCT_IB_MLX5_CMD_OP_QUERY_HCA_CAP);
+    UCT_IB_MLX5DV_SET(query_hca_cap_in, in, op_mod, UCT_IB_MLX5_HCA_CAP_OPMOD_GET_MAX |
+                                                   (UCT_IB_MLX5_CAP_ODP << 1));
+    ret = mlx5dv_devx_general_cmd(md->super.dev.ibv_context, in, sizeof(in),
+                                  out, sizeof(out));
+    if (ret != 0) {
+        return UCS_ERR_IO_ERROR;
+    }
+
+    if (!UCT_IB_MLX5DV_GET(odp_cap, odp, ud_odp_caps.send) ||
+        !UCT_IB_MLX5DV_GET(odp_cap, odp, rc_odp_caps.send) ||
+        !UCT_IB_MLX5DV_GET(odp_cap, odp, rc_odp_caps.write) ||
+        !UCT_IB_MLX5DV_GET(odp_cap, odp, rc_odp_caps.read)) {
+        return UCS_OK;
+    }
+
+    if ((md->super.dev.flags & UCT_IB_DEVICE_FLAG_DC) &&
+        (!UCT_IB_MLX5DV_GET(odp_cap, odp, dc_odp_caps.send) ||
+         !UCT_IB_MLX5DV_GET(odp_cap, odp, dc_odp_caps.write) ||
+         !UCT_IB_MLX5DV_GET(odp_cap, odp, dc_odp_caps.read))) {
+        return UCS_OK;
+    }
+
+    if (md->super.config.odp.max_size == UCS_MEMUNITS_AUTO) {
+        if (UCT_IB_MLX5DV_GET(cmd_hca_cap, cap, umr_extended_translation_offset)) {
+            md->super.config.odp.max_size = 1ul << 55;
+        } else {
+            md->super.config.odp.max_size = 1ul << 28;
+        }
+    }
+
+    if (UCT_IB_MLX5DV_GET(cmd_hca_cap, cap, fixed_buffer_size) &&
+        UCT_IB_MLX5DV_GET(cmd_hca_cap, cap, null_mkey) &&
+        UCT_IB_MLX5DV_GET(cmd_hca_cap, cap, umr_extended_translation_offset)) {
+        md->super.dev.flags |= UCT_IB_DEVICE_FLAG_ODP_IMPLICIT;
```

Comment:
/* assume a devx cabable driver also implements ODP */
maybe also check that ODP for RC,UD is supported in ibv_device_attr_ex?